tests/test_basis.py

```diff
@@ -55,3 +55,51 @@ def bilinf(u, p, v, q, w):
         self.assertTrue('w^2' in y)
 
         self.assertTrue((basis.doflocs[:, D['up'].all()][1] == 1.).all())
+
+
+class TestDoflocs(TestCase):
+
+    mesh = MeshTri
+    elems = [
+        ElementTriP1(),
+        ElementTriRT0(),
+        ElementTriP2(),
+        ElementTriArgyris(),
+    ]
+
+    def runTest(self):
+        """Test that global and local doflocs match in ref. domain."""
+
+        for e in self.elems:
+            self.assertEqual(
```

Comment:
I wonder whether it would be worth making the test both clearer and more rigorous using [`numpy.testing.assert_allclose`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) or similar.

Reply:
Cancellation in the sum of positive and negative errors seems like a possible risk here; e.g. if two doflocs got swapped somehow.

Reply:
Fair enough. Seems that the test is actually failing. Thanks for realizing this.

Reply:
Seems that this test cannot work as such. Unintuitively, the basis functions in reference domains are not mapped one-to-one because of how `Mesh.t2f` is created via `np.unique` and how the global DOF numbering is subsequently created using `Mesh.t2f`.

Changing this would be a fairly large scale change (most of the elements, almost all of the mesh classes, some refinement functions) and it seems to me that there is not any other added benefit than this test would be possible as such.

Basically the numbering of global DOF's is arbitrary and is created in `Mesh._build_dofnum`. Currently this numbering logic simply does not reproduce the numbering that we're expecting for the reference triangle.

I have to think if this would be beneficial in the future or not. For now, I'm inventing another test.

Reply:
I figured out a way to preserve the order of facets but this seemed to broke some assumptions regarding how normal vectors are being mapped from the reference domain.

I'm not going to look into it any further and we should just make notes to ourselves that the global DOF numbering cannot be predicted from the local DOF numbering in any case.

Reply:
I'm now doing some hard thinking on how these things should be tested since there is obviously a possibility of introducing mistakes if the doflocs order does not match with the basis function order.

Reply:
This might be actually a really tough problem in general, because not all elements are nodal (so that `TestNodality` would make sense, e.g., `ElementQuadP` that are based on integrated Legendre polynomials). Do these elements even need doflocs?

Reply:
Earlier I was thinking that doflocs are needed to find dofs on facets, but now that we have `Basis.boundary_dofs` which does it based on the topological entities, and we can first use `Mesh.facets_satisfying` to find the correct entities, I think we don't need those.

Reply:
I now put corresponding `doflocs` rows to `np.nan` if it doesn't make sense to define a location for the particular DOF.